### PR TITLE
[REF] sale_purchase: reorganize with hooks

### PR DESCRIPTION
**Pull request by JKE-be**
Add a lot of hook to allow a custom module to reuse the standard sale_purchase module.

### CODE REVIEW

**Comment by JKE-be** [comment]
@robodoo delegate=madprog

File: addons/sale_purchase/models/sale_order_line.py [hunk 1]
```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
+        if warning and not suppliers:
+            raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (self.product_id.display_name,))
+        return suppliers
+
+    def _purchase_service_match_purchase_order(self, partner, company=False):
+        return self.env['purchase.order'].search([
+            ('partner_id', '=', partner.id),
+            ('state', '=', 'draft'),
+            ('company_id', '=', (company and company or self.env.company).id),
+        ], order='id desc')
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:221]: The extracted code did not specify a sorting order, so this could change the behavior.
It should be moved to a separate commit.
```suggestion
        ], limit=1)
```

File: addons/sale_purchase/models/sale_order_line.py [hunk 2]
```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
+
+    def _create_purchase_order(self, supplierinfo):
+        values = self._purchase_service_prepare_order_values(supplierinfo)
+        return self.env['purchase.order'].with_context(mail_create_nosubscribe=True).create(values)
+
+    def _match_or_create_purchase_order(self, supplierinfo):
+        purchase_order = self._purchase_service_match_purchase_order(supplierinfo.partner_id)[:1]
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:228]: If we restore the `limit=1` in `_purchase_service_match_purchase_order`, there is no need to limit the output to the first item as the length should already be 0 or 1.
```suggestion
        purchase_order = self._purchase_service_match_purchase_order(supplierinfo.partner_id)
```

File: addons/sale_purchase/models/sale_order_line.py [hunk 3]
```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
-                    origins = purchase_order.origin.split(', ') + origins
+                origins = (purchase_order.origin or '').split(', ')
                 if so_name not in origins:
-                    origins += [so_name]
-                    purchase_order.write({
-                        'origin': ', '.join(origins)
-                    })
-            supplier_po_map[partner_supplier.id] = purchase_order
+                    purchase_order.write({'origin': ', '.join(origins + [so_name])})
+                supplier_po_map[partner_supplier.id] = purchase_order
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:255]: If the PO was created on line 249, it should still be added to the map, so we need to put it outside of the if/else block (as in the extracted code):
```suggestion
            supplier_po_map[partner_supplier.id] = purchase_order
```

File: addons/sale_purchase/models/sale_order_line.py [hunk 4]
```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
             # add a PO line to the PO
             values = line._purchase_service_prepare_line_values(purchase_order, quantity=quantity)
-            purchase_line = line.env['purchase.order.line'].create(values)
+            purchase_lines |= line.env['purchase.order.line'].create(values)
 
-            # link the generated purchase to the SO line
-            sale_line_purchase_map.setdefault(line, line.env['purchase.order.line'])
-            sale_line_purchase_map[line] |= purchase_line
-        return sale_line_purchase_map
+        return purchase_lines
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:261]: Here the signature of the method has changed, as we return a recordset instead of a dictionary SO line => PO line.
Maybe this result is not used anywhere in our code, but it could be used by third-party modules.

File: addons/sale_purchase/models/sale_order_line.py [hunk 5]
```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
         """ Create a Purchase for the first time from the sale line. If the SO line already created a PO, it
             will not create a second one.
         """
-        sale_line_purchase_map = {}
         for line in self:
             # Do not regenerate PO line if the SO line has already created one in the past (SO cancel/reconfirmation case)
             if line.product_id.service_to_purchase and not line.purchase_line_count:
-                result = line._purchase_service_create()
-                sale_line_purchase_map.update(result)
-        return sale_line_purchase_map
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:257]: We should not change the method signature in a frozen version.

**Review by madprog — CHANGES_REQUESTED** [review]
Hello and thank you for this PR that will allow us to plug our customization into this module!

However I have seen a few things to change. Some of them are related to the stable policy, because the signature of existing methods is being changed, and others because, according to the commit's message, it only should extract some code to methods, and change it as little as possible to avoid any disruption. I think that other changes belong to separate commits, or to the master branch.

File: addons/sale_purchase/models/sale_order_line.py (hunk 3 above)
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:255]: not sure, if we create it, we already know that we have the origin on the PO (purpose of the else part).

So the second time, the match will return the correct purchase order, and avoid to resplit/join the origin without modification :thinking: but `match` vs `split/join` not sure what is the worst. All this code is not optimized and could be rewrite in master.
I will unindent it in stable, since it was not obvious ...

File: addons/sale_purchase/models/sale_order_line.py [hunk 6]
```diff
@@ -137,6 +138,39 @@ def _purchase_service_prepare_order_values(self, supplierinfo):
             'fiscal_position_id': fpos.id,
         }
 
+    def _purchase_service_get_price_unit_and_taxes(self, supplierinfo, purchase_order):
+        supplier_taxes = self.product_id.supplier_taxes_id.filtered(lambda t: t.company_id == purchase_order.company_id)
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:142]: As discuss in PM, but for history...

It should be more 'correct' since we apply taxes from the company of the PO company and not the current user that could differ.
In our case, it will allow as customization to assign specific company to some PO whatever the company of the current user / sale order.

File: addons/sale_purchase/models/sale_order_line.py [hunk 7]
```diff
@@ -137,6 +138,39 @@ def _purchase_service_prepare_order_values(self, supplierinfo):
...
+            price_unit = self.env['account.tax'].sudo()._fix_tax_included_price_company(supplierinfo.price, supplier_taxes, taxes, purchase_order.company_id)
+            if purchase_order.currency_id and supplierinfo.currency_id != purchase_order.currency_id:
+                price_unit = supplierinfo.currency_id._convert(
+                    price_unit,
+                    purchase_order.currency_id,
+                    purchase_order.company_id,
+                    fields.Date.context_today(self)
+                )
+        else:
+            price_unit = 0
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:154]: nitpicking, but no cost to change it.

File: addons/sale_purchase/models/sale_order_line.py [hunk 8]
```diff
@@ -145,10 +179,7 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
         """
         self.ensure_one()
         # compute quantity from SO line UoM
-        product_quantity = self.product_uom_qty
-        if quantity:
-            product_quantity = quantity
-
+        product_quantity = quantity if quantity else self.product_uom_qty
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:182]: reverted to minimal diff

File: addons/sale_purchase/models/sale_order_line.py [hunk 9]
```diff
@@ -103,7 +101,7 @@ def _purchase_increase_ordered_qty(self, new_qty, origin_values):
             :param origin_values: map from sale line id to old value for the ordered quantity (dict)
         """
         for line in self:
-            last_purchase_line = self.env['purchase.order.line'].search([('sale_line_id', '=', line.id)], order='create_date DESC', limit=1)
+            last_purchase_line = self.purchase_line_ids[:1]
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:104]: reverted to minimal diff

File: addons/sale_purchase/models/sale_order_line.py [hunk 10]
```diff
@@ -49,10 +49,9 @@ def create(self, values):
         return lines
 
     def write(self, values):
-        increased_lines = None
-        decreased_lines = None
-        increased_values = {}
-        decreased_values = {}
+        increased_lines = decreased_lines = None
+        increased_values = decreased_values = {}
+
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:54]: reverted to minimal diff

File: addons/sale_purchase/models/sale_order.py [hunk 11]
```diff
@@ -72,4 +72,5 @@ def _activity_cancel_on_purchase(self):
                 render_context={
                     'sale_orders': sale_order_lines.mapped('order_id'),
                     'sale_order_lines': sale_order_lines,
-            })
+                }
+            )
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order.py:76]: reverted to minimal diff

File: addons/sale_purchase/models/sale_order_line.py [hunk 12]
```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
             'order_id': purchase_order.id,
             'sale_line_id': self.id,
         }
 
+    def _purchase_service_match_supplier(self, warning=True):
+        # determine vendor of the order (take the first matching company and product)
+        suppliers = self.product_id._select_seller(quantity=self.product_uom_qty, uom_id=self.product_uom)
+        if warning and not suppliers:
+            raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (self.product_id.display_name,))
+        return suppliers
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:214]: revert to minimal diff, but _select_seller always return max 1 supplier

File: addons/sale_purchase/models/sale_order_line.py (hunk 5 above)
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:257]: reverted to minimal diff, but a lot of useless code...

File: addons/sale_purchase/models/sale_order_line.py (hunk 4 above)
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:261]: reverted to minimal diff, but a lot of useless code...
As discuss, it was in un-deployed saas version, so the risk of override of this function is really weak

File: addons/sale_purchase/models/sale_order_line.py (hunk 2 above)
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:228]: moved to second commit

File: addons/sale_purchase/models/sale_order_line.py [hunk 13]
```diff
@@ -137,6 +138,39 @@ def _purchase_service_prepare_order_values(self, supplierinfo):
             'fiscal_position_id': fpos.id,
         }
 
+    def _purchase_service_get_price_unit_and_taxes(self, supplierinfo, purchase_order):
+        supplier_taxes = self.product_id.supplier_taxes_id.filtered(lambda t: t.company_id == purchase_order.company_id)
+        taxes = purchase_order.fiscal_position_id.map_tax(supplier_taxes)
+        if supplierinfo:
+            price_unit = self.env['account.tax'].sudo()._fix_tax_included_price_company(supplierinfo.price, supplier_taxes, taxes, purchase_order.company_id)
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:147]: moved to second commit

File: addons/sale_purchase/models/sale_order_line.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:144]: moved to second commit

File: addons/sale_purchase/models/sale_order_line.py [hunk 14]
```diff
@@ -137,6 +138,39 @@ def _purchase_service_prepare_order_values(self, supplierinfo):
...
+            'company_id': purchase_order.company_id.id,
+        }
+        if supplierinfo:
+            product_ctx.update({'seller_id': supplierinfo.id})
+        else:
+            product_ctx.update({'partner_id': purchase_order.partner_id.id})
+        product = self.product_id.with_context(**product_ctx)
+        name = product.display_name
+        if quantity < 0:
+            name = _("Refund of %r") % name
```
[the remark this answers is not in the document]
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py]: moved in internal, but will be a good imp in code base too

File: addons/sale_purchase/models/sale_order_line.py [hunk 15]
```diff
@@ -159,80 +193,69 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
...
     def _purchase_service_create(self, quantity=False):
         """ On Sales Order confirmation, some lines (services ones) can create a purchase order line and maybe a purchase order.
             If a line should create a RFQ, it will check for existing PO. If no one is find, the SO line will create one, then adds
             a new PO line. The created purchase order line will be linked to the SO line.
             :param quantity: the quantity to force on the PO line, expressed in SO line UoM
         """
-        PurchaseOrder = self.env['purchase.order']
         supplier_po_map = {}
         sale_line_purchase_map = {}
+        purchase_lines = self.env['purchase.order.line']
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:244]: In first commit, `return purchase_lines` is the only usage of this variable and it is dead code (after a `return` statement).
```suggestion
```

File: addons/sale_purchase/models/sale_order_line.py [hunk 16]
```diff
@@ -244,6 +267,9 @@ def _purchase_service_create(self, quantity=False):
             sale_line_purchase_map[line] |= purchase_line
         return sale_line_purchase_map
 
+
+        return purchase_lines
+
```
Comment (madprog) [inline addons/sale_purchase/models/sale_order_line.py:272]: This is dead code in the first commit, as it is after `return sale_line_purchase_map`
```suggestion
```

**Review by madprog — CHANGES_REQUESTED** [review]

File: addons/sale_purchase/models/sale_order_line.py (hunk 16 above)
Comment (JKE-be) [reply addons/sale_purchase/models/sale_order_line.py:272]: oupsy, probably a rebase failed

**Review by madprog — APPROVED** [review]

**Comment by madprog** [comment]
@robodoo r+

**Comment by madprog** [comment]
@robodoo rebase-ff
